src/main/scala/rise/core/uniqueNames.scala

```diff
@@ -1,34 +1,28 @@
 package rise.core
 
+import util.monads._
 import rise.core.traverse._
 import rise.core.types._
 
 object uniqueNames {
-  private case class CountingVisitor(
-    var values: Map[Identifier, Int],
-    var types: Map[Kind.Identifier, Int]
-  ) extends PureTraversal {
-    override def identifier[I <: Identifier]: VarType => I => Pure[I] = {
-      case Binding => i =>
-        values = values + (i -> (values.getOrElse(i, 0) + 1))
-        return_(i)
+  private val CountingVisitor = new PairMonoidTraversal[(Seq[Identifier], Seq[Kind.Identifier]), Pure] {
```

Comment:
this is not a `CountingVisitor` anymore I would say?

Reply:
Fixed!